crates/datasources/src/bigquery/mod.rs

```diff
@@ -476,37 +477,55 @@ fn bigquery_table_to_arrow_schema(table: &Table) -> Result<ArrowSchema> {
-            FieldType::Datetime => DataType::Timestamp(TimeUnit::Nanosecond, None),
-            FieldType::Timestamp => DataType::Timestamp(TimeUnit::Nanosecond, Some("UTC".into())),
-            FieldType::Time => DataType::Time64(TimeUnit::Nanosecond),
-            FieldType::Numeric => DataType::Decimal128(38, 9),
-            FieldType::Geography => DataType::Utf8,
-            other => return Err(BigQueryError::UnsupportedBigQueryType(other.clone())),
-        };
-
-        let field = Field::new(&field.name, arrow_typ, true);
-        arrow_fields.push(field);
+        let resolved_field = table_field_schema_to_arrow_datatype(field);
+        if let Ok(resolved_field) = resolved_field {
+            arrow_fields.push(resolved_field);
+        }
     }
-
     Ok(ArrowSchema::new(arrow_fields))
 }
 
+fn table_field_schema_to_arrow_datatype(field: &BigQuerySchema) -> Result<Field> {
+    // See <https://cloud.google.com/bigquery/docs/reference/storage#arrow_schema_details>
+    // for how BigQuery types map to Arrow types.
+    let arrow_typ = match &field.r#type {
+        FieldType::Bool | FieldType::Boolean => DataType::Boolean,
+        FieldType::String => DataType::Utf8,
+        FieldType::Integer | FieldType::Int64 => DataType::Int64,
+        FieldType::Float | FieldType::Float64 => DataType::Float64,
+        FieldType::Bytes => DataType::Binary,
+        FieldType::Date => DataType::Date32,
+        // BigQuery actually returns times with microsecond precision. We
+        // aim to work only with nanoseconds to have uniformity accross the
+        // codebase. It's also easier to have interop with datafusion since
+        // with many things like type inference datafusion uses nanosecond.
+        // This cast is done when the stream is received by using the
+        // `crate::common::util::normalize_batch` function.
+        FieldType::Datetime => DataType::Timestamp(TimeUnit::Nanosecond, None),
+        FieldType::Timestamp => DataType::Timestamp(TimeUnit::Nanosecond, Some("UTC".into())),
+        FieldType::Time => DataType::Time64(TimeUnit::Nanosecond),
+        FieldType::Numeric => DataType::Decimal128(38, 9),
+        FieldType::Bignumeric => DataType::Decimal256(76, 9),
+        FieldType::Geography => DataType::Utf8,
+        FieldType::Record | FieldType::Struct => {
+            let mut record: Vec<Field> = Vec::new();
+            let fields = field.fields.clone();
+            if let Some(fields) = fields {
+                for field in fields.iter() {
+                    let resolved_field = table_field_schema_to_arrow_datatype(field);
+                    if let Ok(resolved_field) = resolved_field {
+                        record.push(resolved_field);
+                    }
+                }
+            }
+            DataType::Struct(Fields::from(record))
+        }
+        other => return Err(BigQueryError::UnsupportedBigQueryType(other.clone())),
```

Comment:
I think the last type we'd need to support is array. Would that be something we could throw in?

Reply:
The only unsupported type left is Json